lib/broadway/producer.ex

```diff
@@ -236,4 +277,27 @@ defmodule Broadway.Producer do
             "struct. In case you're using a standard GenStage producer, please set the " <>
             ":transformer option to transform produced events into message structs"
   end
+
+  defp maybe_rate_limit_and_buffer_messages(state, messages) do
+    if state[:rate_limiting_enabled?] && messages != [] do
+      rate_limit_and_buffer_messages(state, messages)
+    else
+      {state, messages}
+    end
+  end
+
+  defp rate_limit_and_buffer_messages(state, messages) do
+    message_count = length(messages)
+    allowed = state.rate_limiting_allowed
+
+    if allowed >= message_count do
+      state = update_in(state.rate_limiting_allowed, &(&1 - message_count))
+      {state, messages}
+    else
+      {messages_to_send, messages_to_buffer} = Enum.split(messages, allowed)
```

Comment:
I generally have a function called `split_wth_count`, that does the splitting and counting at the same time, to avoid multiple traversals.